File: src/openmap/src/main/java/com/bbn/openmap/layer/link/LinkRectangle.java
// **********************************************************************
// 
// <copyright>
// 
//  BBN Technologies
//  10 Moulton Street
//  Cambridge, MA 02138
//  (617) 873-8000
// 
//  Copyright (C) BBNT Solutions LLC. All rights reserved.
// 
// </copyright>
// **********************************************************************
// 
// $Source: /cvs/distapps/openmap/src/openmap/com/bbn/openmap/layer/link/LinkRectangle.java,v $
// $RCSfile: LinkRectangle.java,v $
// $Revision: 1.6 $
// $Date: 2009/01/21 01:24:41 $
// $Author: dietrick $
// 
// **********************************************************************

package com.bbn.openmap.layer.link;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

import com.bbn.openmap.omGraphics.OMRect;
import com.bbn.openmap.util.Debug;

/**
 * Read and write the Link protocol for rectangles.
 */
public class LinkRectangle implements LinkGraphicConstants,
        LinkPropertiesConstants {

    /**
     * Create a lat/lon rectangle.
     * 
     * @param lt1 latitude of north edge, decimal degrees.
     * @param ln1 longitude of west edge, decimal degrees.
     * @param lt2 latitude of south edge, decimal degrees.
     * @param ln2 longitude of east edge, decimal degrees.
     * @param lType line type - see lineType.
     * @param properties description of drawing attributes.
     * @param dos DataOutputStream
     * @throws IOException
     */
    public static void write(float lt1, float ln1, float lt2, float ln2,
                             int lType, LinkProperties properties,
                             DataOutputStream dos) throws IOException {
        LinkRectangle.write(lt1, ln1, lt2, ln2, lType, -1, properties, dos);
    }

    /**
     * Create a lat/lon rectangle.
     * 
     * @param lt1 latitude of north edge, decimal degrees.
     * @param ln1 longitude of west edge, decimal degrees.
     * @param lt2 latitude of south edge, decimal degrees.
     * @param ln2 longitude of east edge, decimal degrees.
     * @param lType line type - see lineType.
     * @param nsegs number of segment points (only for LINETYPE_GREATCIRCLE or
     *        LINETYPE_RHUMB line types, and if &lt; 1, this value is generated
     *        internally)
     * @param properties description of drawing attributes.
     * @param dos DataOutputStream
     * @throws IOException
     */
    public static void write(float lt1, float ln1, float lt2, float ln2,
                             int lType, int nsegs, LinkProperties properties,
                             DataOutputStream dos) throws IOException {

        dos.write(Link.RECTANGLE_HEADER.getBytes());
        dos.writeByte(GRAPHICTYPE_RECTANGLE);
        dos.writeByte(RENDERTYPE_LATLON);
        dos.writeByte(lType);
        dos.writeFloat(lt1);
        dos.writeFloat(ln1);
        dos.writeFloat(lt2);
        dos.writeFloat(ln2);

        dos.writeInt(nsegs);
        properties.write(dos);
    }

    /**
     * Construct an XY rectangle. It doesn't matter which corners of the
     * rectangle are used, as long as they are opposite from each other.
     * 
     * @param px1 x pixel position of the first corner relative to the window
     *        origin
     * @param py1 y pixel position of the first corner relative to the window
     *        origin
     * @param px2 x pixel position of the second corner relative to the window
     *        origin
     * @param py2 y pixel position of the second corner relative to the window
     *        origin
     * @param properties description of drawing attributes.
     * @param dos DataOutputStream
     * @throws IOException
     */
    public static void write(int px1, int py1, int px2, int py2,
                             LinkProperties properties, DataOutputStream dos)
            throws IOException {

        dos.write(Link.RECTANGLE_HEADER.getBytes());
        dos.writeByte(GRAPHICTYPE_RECTANGLE);
        dos.writeByte(RENDERTYPE_XY);
        dos.writeInt(px1);
        dos.writeInt(py1);
        dos.writeInt(px2);
        dos.writeInt(py2);
        properties.write(dos);
    }

    /**
     * Construct an XY rectangle relative to a lat/lon point
     * (RENDERTYPE_OFFSET). It doesn't matter which corners of the rectangle are
     * used, as long as they are opposite from each other.
     * 
     * @param lt1 latitude of the reference point, decimal degrees.
     * @param ln1 longitude of the reference point, decimal degrees.
     * @param px1 x pixel position of the first corner relative to the reference
     *        point
     * @param py1 y pixel position of the first corner relative to the reference
     *        point
     * @param px2 x pixel position of the second corner relative to the
     *        reference point
     * @param py2 y pixel position of the second corner relative to the
     *        reference point
     * @param properties description of drawing attributes.
     * @param dos DataOutputStream
     * @throws IOException
     */
    public static void write(float lt1, float ln1, int px1, int py1, int px2,
                             int py2, LinkProperties properties,
                             DataOutputStream dos) throws IOException {

        dos.write(Link.RECTANGLE_HEADER.getBytes());
        dos.writeByte(GRAPHICTYPE_RECTANGLE);
        dos.writeByte(RENDERTYPE_OFFSET);
        dos.writeFloat(lt1);
        dos.writeFloat(ln1);
        dos.writeInt(px1);
        dos.writeInt(py1);
        dos.writeInt(px2);
        dos.writeInt(py2);
        properties.write(dos);
    }

    /**
     * Write an OMRect to the link.
     */
    public static void write(OMRect rect, Link link, LinkProperties props)
            throws IOException {

        switch (rect.getRenderType()) {
        case OMRect.RENDERTYPE_LATLON:
            LinkRectangle.write((float) rect.getNorthLat(),
                    (float) rect.getWestLon(),
                    (float) rect.getSouthLat(),
                    (float) rect.getEastLon(),
                    rect.getLineType(),
                    props,
                    link.dos);
            break;
        case OMRect.RENDERTYPE_XY:
            LinkRectangle.write(rect.getLeft(),
                    rect.getTop(),
                    rect.getRight(),
                    rect.getBottom(),
                    props,
                    link.dos);
            break;
        case OMRect.RENDERTYPE_OFFSET:
            LinkRectangle.write((float) rect.getNorthLat(),
                    (float) rect.getWestLon(),
                    rect.getLeft(),
                    rect.getTop(),
                    rect.getRight(),
                    rect.getBottom(),
                    props,
                    link.dos);
            break;
        default:
            Debug.error("LinkRectangle.write: rect rendertype unknown.");
        }
    }

    /**
     * Read the DataInputStream, and create an OMRect. Assumes that the
     * LinkRectangle header has been read from the link.
     * 
     * @param dis DataInputStream
     * @return OMRect
     * @throws IOException
     * @see com.bbn.openmap.omGraphics.OMRect
     */
    public static OMRect read(DataInputStream dis) throws IOException {
        return read(dis, null);
    }

    /**
     * Read the DataInputStream, and create an OMRect. Assumes that the
     * LinkRectangle header has been read from the link.
     * 
     * @param dis DataInputStream
     * @param propertiesBuffer a LinkProperties object used to cache previous
     *        settings that can be set on the OMRect being read.
     * @return OMRect
     * @throws IOException
     * @see com.bbn.openmap.omGraphics.OMRect
     */
    public static OMRect read(DataInputStream dis,
                              LinkProperties propertiesBuffer)
            throws IOException {
        OMRect rect = null;
        int x1, y1, x2, y2;
        double lt1, ln1, lt2, ln2;

        int renderType = dis.readByte();

        switch (renderType) {
        case RENDERTYPE_LATLON:
            int lineType = dis.readByte();
            lt1 = dis.readFloat();
            ln1 = dis.readFloat();
            lt2 = dis.readFloat();
            ln2 = dis.readFloat();
            int nsegs = dis.readInt();

            rect = new OMRect(lt1, ln1, lt2, ln2, lineType, nsegs);
            break;
        case RENDERTYPE_XY:
            x1 = dis.readInt();
            y1 = dis.readInt();
            x2 = dis.readInt();
            y2 = dis.readInt();

            rect = new OMRect(x1, y1, x2, y2);
            break;
        case RENDERTYPE_OFFSET:
            lt1 = dis.readFloat();
            ln1 = dis.readFloat();

            x1 = dis.readInt();
            y1 = dis.readInt();
            x2 = dis.readInt();
            y2 = dis.readInt();

            rect = new OMRect(lt1, ln1, x1, y1, x2, y2);
            break;
        default:
        }

        if (rect != null) {
            LinkProperties.loadPropertiesIntoOMGraphic(dis,
                    rect,
                    propertiesBuffer);
        }

        return rect;
    }

}

File: src/cserver/link/src/LinkRectangle.c
/* **********************************************************************
 * 
 * <copyright>
 * 
 *  BBN Technologies, a Verizon Company
 *  10 Moulton Street
 *  Cambridge, MA 02138
 *  (617) 873-8000
 * 
 *  Copyright (C) BBNT Solutions LLC. All rights reserved.
 * 
 * </copyright>
 * **********************************************************************
 * 
 * $Source: /cvs/distapps/openmap/src/cserver/link/src/LinkRectangle.c,v $
 * $RCSfile: LinkRectangle.c,v $
 * $Revision: 1.2 $
 * $Date: 2004/01/26 19:07:09 $
 * $Author: dietrick $
 * 
 * **********************************************************************
 */

#include <stdlib.h>
#include <stdio.h>

#include "LinkRectangle.h"
#include "Link.h"
#include "GlobalConstants.h"
#include "Response.h"

int WriteRectangleHeader(LinkSocket *linkSocket)
{
    int check = OK;
    check = check || WriteChars(linkSocket, RECTANGLE_HEADER,
                                lRECTANGLE_HEADER);
    check = check || WriteInteger(linkSocket, GRAPHICTYPE_RECTANGLE);
    return check;
}

int WriteLinkRectangleLatLon(LinkSocket *linkSocket, int ltype,
                             double NWlat, double NWlon, /*NW = North West*/
                             double SElat, double SElon, /*SE = South East*/
                             int nsegs,
                             LinkArgs *linkArgs)
{
    int check = OK;
    check = check || WriteRectangleHeader(linkSocket);
    check = check || WriteInteger(linkSocket, RENDERTYPE_LATLON);
    check = check || WriteInteger(linkSocket, ltype);
    check = check || WriteFloat(linkSocket,(float)NWlat);
    check = check || WriteFloat(linkSocket,(float)NWlon);
    check = check || WriteFloat(linkSocket,(float)SElat);
    check = check || WriteFloat(linkSocket,(float)SElon);
    check = check || WriteInteger(linkSocket, nsegs);
    check = check || WriteLinkArgs(linkSocket, linkArgs);

    return check;
    /* Returns -1 if there was a memory allocation error in WriteLinkArgs() */
}


int WriteLinkRectangleXY(LinkSocket *linkSocket,
                         int ulx, int uly, 
                         int lrx, int lry,
                         LinkArgs *linkArgs)
{
    int check = OK;
    check = check || WriteRectangleHeader(linkSocket);
    check = check || WriteInteger(linkSocket, RENDERTYPE_XY);
    check = check || WriteInteger(linkSocket, ulx);
    check = check || WriteInteger(linkSocket, uly);
    check = check || WriteInteger(linkSocket, lrx);
    check = check || WriteInteger(linkSocket, lry);  
    check = check || WriteLinkArgs(linkSocket, linkArgs);

    return check;
    /* Returns -1 if there was a memory allocation error in WriteLinkArgs() */
}


int WriteLinkRectangleOffset(LinkSocket *linkSocket,
                             double lat, double lon,
                             int ulx, int uly,
                             int lrx, int lry,                     
                             LinkArgs *linkArgs)
{
    int check = OK;
    check = check || WriteRectangleHeader(linkSocket);
    check = check || WriteInteger(linkSocket, RENDERTYPE_OFFSET);
    check = check || WriteFloat(linkSocket,(float)lat);
    check = check || WriteFloat(linkSocket,(float)lon);
    check = check || WriteInteger(linkSocket, ulx);
    check = check || WriteInteger(linkSocket, uly);
    check = check || WriteInteger(linkSocket, lrx);
    check = check || WriteInteger(linkSocket, lry);
    check = check || WriteLinkArgs(linkSocket, linkArgs);

    return check;
    /* Returns -1 if there was a memory allocation error in WriteLinkArgs() */
}

int BufferedWriteRectangleHeader(char *toBuffer)
{
    int byteswritten = 0;
    byteswritten += BufferedWriteChars(&toBuffer[byteswritten],
                                       RECTANGLE_HEADER,
                                       lRECTANGLE_HEADER);
    byteswritten += BufferedWriteInteger(&toBuffer[byteswritten],
                                         GRAPHICTYPE_RECTANGLE);
    return byteswritten;
}


int BufferedWriteLinkRectangleLatLon(LinkSocket *linkSocket, int ltype,
                                     double NWlat, double NWlon, /*NW = North West*/
                                     double SElat, double SElon, /*SE = South East*/
                                     int nsegs,
                                     LinkArgs *linkArgs)
{
    int buffercount;
    char *buffer;
    int retval;
    /*
      buffercount corresponds to data being written in bytes below.
    */
    buffercount = lRECTANGLE_HEADER +  /* Bytes used by Header*/
        N_BYTES_PER_INTEGER +     /*Bytes used by GRAPHIC_TYPE. It is an Integer*/
        N_BYTES_PER_INTEGER +     /*Bytes used by RENDERTYPE identifier*/
        4*N_BYTES_PER_FLOAT +     /*Bytes used by lat and lon*/
        2*N_BYTES_PER_INTEGER +   /*Bytes used by ltype and nsegs*/
        LinkSizeOfLinkArgs(linkArgs);
    
    buffer = (char *)malloc(buffercount);
    if (buffer == NULL)
      return -1; /* Memory allocation error */

    buffercount=0;
    
    buffercount += BufferedWriteRectangleHeader(&buffer[buffercount]);
    buffercount += BufferedWriteInteger(&buffer[buffercount], RENDERTYPE_LATLON);
    buffercount += BufferedWriteInteger(&buffer[buffercount], ltype);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)NWlat);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)NWlon);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)SElat);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)SElon);
    buffercount += BufferedWriteInteger(&buffer[buffercount], nsegs);
    buffercount += BufferedWriteLinkArgs(&buffer[buffercount], linkArgs);

    retval = WriteChars(linkSocket, buffer,buffercount);         
    free(buffer);
    return retval;
}


int BufferedWriteLinkRectangleXY(LinkSocket *linkSocket,
                                 int ulx, int uly, 
                                 int lrx, int lry,
                                 LinkArgs *linkArgs)
{

    int buffercount;
    char *buffer;
    int retval;
    /*
      buffercount corresponds to data being written in bytes below.
    */
    buffercount = lRECTANGLE_HEADER +  /* Bytes used by Header*/
        N_BYTES_PER_INTEGER +     /*Bytes used by GRAPHIC_TYPE. It is an Integer*/
        N_BYTES_PER_INTEGER +     /*Bytes used by RENDERTYPE identifier*/
        4*N_BYTES_PER_INTEGER +   /*Bytes used by bounding co-ordinates of rectangle*/
        LinkSizeOfLinkArgs(linkArgs);
    
    buffer = (char *)malloc(buffercount);
    if (buffer == NULL)
      return -1; /* Memory allocation error */

    buffercount=0;
    
    buffercount += BufferedWriteRectangleHeader(&buffer[buffercount]);
    buffercount += BufferedWriteInteger(&buffer[buffercount], RENDERTYPE_XY);
    buffercount += BufferedWriteInteger(&buffer[buffercount], ulx);
    buffercount += BufferedWriteInteger(&buffer[buffercount], uly);
    buffercount += BufferedWriteInteger(&buffer[buffercount], lrx);
    buffercount += BufferedWriteInteger(&buffer[buffercount], lry);  
    buffercount += BufferedWriteLinkArgs(&buffer[buffercount], linkArgs);

    retval = WriteChars(linkSocket, buffer,buffercount);         
    free(buffer);
    return retval;
}


int BufferedWriteLinkRectangleOffset(LinkSocket *linkSocket,
                                     double lat, double lon,
                                     int ulx, int uly,
                                     int lrx, int lry,                     
                                     LinkArgs *linkArgs)
{

    int buffercount;
    char *buffer;
    int retval;
    /*
      buffercount corresponds to data being written in bytes below.
    */
    buffercount = lRECTANGLE_HEADER +  /* Bytes used by Header*/
        N_BYTES_PER_INTEGER +     /*Bytes used by GRAPHIC_TYPE. It is an Integer*/
        N_BYTES_PER_INTEGER +     /*Bytes used by RENDERTYPE identifier*/
        2*N_BYTES_PER_FLOAT +     /*Bytes used by lat and lon*/
        4*N_BYTES_PER_INTEGER +   /*Bytes used by bounding co-ordinates of rectangle*/
        LinkSizeOfLinkArgs(linkArgs);
               

    buffer = (char *)malloc(buffercount);
    if (buffer == NULL)
      return -1; /* Memory allocation error */

    buffercount=0;

    buffercount += BufferedWriteRectangleHeader(&buffer[buffercount]);
    buffercount += BufferedWriteInteger(&buffer[buffercount], RENDERTYPE_OFFSET);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)lat);
    buffercount += BufferedWriteFloat(&buffer[buffercount],(float)lon);
    buffercount += BufferedWriteInteger(&buffer[buffercount], ulx);
    buffercount += BufferedWriteInteger(&buffer[buffercount], uly);
    buffercount += BufferedWriteInteger(&buffer[buffercount], lrx);
    buffercount += BufferedWriteInteger(&buffer[buffercount], lry);
    buffercount += BufferedWriteLinkArgs(&buffer[buffercount], linkArgs);

    retval = WriteChars(linkSocket, buffer,buffercount);         
    free(buffer);
    return retval;
}
